### ISSUE TITLE
Provide an approved way to subclass DNA

### ISSUE THREAD

**Issue by clintval** [scikit-bio/scikit-bio#1637]
I would like to use [`skbio.DNA`](http://scikit-bio.org/docs/latest/generated/skbio.sequence.DNA.html) in my library by subclassing it to provide new functionality upon the class without mutating any of its original construction. I understand the desire to protect the class, but that limits my ability to build with your library!

```
Notes
-----
Subclassing is disabled for DNA, because subclassing makes
it possible to change the alphabet, and certain methods rely on the
IUPAC alphabet. If a custom sequence alphabet is needed, inherit directly
from ``GrammaredSequence``.
```

Would you consider either:

1. Removing the [`DisableSubclassingMeta`](https://github.com/biocore/scikit-bio/blob/de756f06c3e63448f129bf39ad39c8e159211ae2/skbio/sequence/_grammared_sequence.py#L84-L93)
2. Providing an easily inheritable version of the class

Python once helped the developer subclass `dict` by providing `collection.UserDict` although that's not used as much now that [new-style classes](https://www.python.org/doc/newstyle/) exist.

Currently, if I want to use `skbio.DNA` with additional methods, I must copy-and-paste much of what you have already developed which increases my testing burden and causes me to miss out on upstream enhancements and patches.

**Comment by xguse** [comment]
I am going to leave this here in case it helps someone else. The decision to make the Seq classes subclassable is baffling to me. If I want to add a SINGLE method like `mutate` or `to_fasta` I have to subclass a class that gives me NOTHING resembling a `DNA` class?  So I have to subclass the arcane MetaClass and **REIMPLEMENT** `DNA` by myself before I can add a simple method? WTF?

But its where we are. 

Dear reader in the future: I have resorted to blatant monkey-patching. It works. Its ugly and I feel dirty doing it (but not as dirty as choosing to use a metaclass as a baseclass... ;p). 

But it works.

They forced us after all.

**Comment by xguse** [comment]
> Currently, if I want to use `skbio.DNA` with additional methods, I must copy-and-paste much of what you have already developed which increases my testing burden and causes me to miss out on upstream enhancements and patches.

@qiyunzhu This is my main problem, expressed much more succinctly and with less frustration than I was able manage, I am ashamed to say.

I have built a `Mutation` class. I have written a `apply_mutations` method as well as a `to_fasta` method that I would like to add to `DNA` and `Protein` in order to teach them how to mutate themselves into a new seq and use the metadata stored in them to dump fasta records of themselves. They are simple and would help me out a lot. But I am not going to spend the time implementing the base features of your main Sequence classes just to have them eventually diverge from your main classes as time goes on. It is for this reason thatI resorted to monkey patching instead. But it is not optimal.

Forcing people to inherit from a bare bones class just seems to defeat the purpose of inheritance. I understand that people can abuse the class if they are not careful in their subclassing, but... I mean... isnt that their own fault?

Anyway, I would REALLY like to continue adapting your classes but it really made me consider leaving the library for the bloated and exceedingly unpythonic (in my opinion) BioPython (shutter).

I didnt, but I may have too later.

**Comment by qiyunzhu** [comment]
@xguse Thanks for explaining your thoughts!
@mortonjt and I discussed and we both think it is a good suggestion. It unlocks scikit-bio's potential for secondary development and to support a large spectrum of downstream applications. @wasade and @gregcaporaso do you have specific concerns on this?

**Comment by qiyunzhu** [comment]
If keeping the alphabet immutable is important (although I don't see why), we may adopt the new trick: `@dataclass(frozen=True)` according to this [post](https://stackoverflow.com/a/58992808). Just tested and working.

**Comment by ebolyen** [comment]
Wanting to subclass this does make sense, especially for things like adding methods. The original concern as best I can recall was indeed just maintaining invariant on the alphabet. Starting from scratch with GrammeredSequence is indeed quite tedius. I like @qiyunzhu's approach of freezing the alphabet (and related mappings) which would enable the kind of class extensions being discussed, without causing aligners or DNA sequence methods from failing because an IUPAC code has been swapped or added.

We also knew of situations where certain class invariants in Pycogent had been replaced on the class, which unintentionally persisted beyond the scope where those replacements made sense. Causing issues downstream, as sys.modules caches everything. So that's also going to be some of the conservative approach you'll see throughout scikit-bio. Freezing the alphabet and preventing class mutation on those specific attributes I think strikes a good balance.

**Comment by qiyunzhu** [comment]
@ebolyen Thanks for sharing your opinion! I tested a bit and couldn't figure how a user can modify IUPAC codes in the current scikit-bio. Do you know how? Here I got:

```
>>> from skbio import DNA
>>> seq = DNA('ACCGAAT')
>>> seq.degenerate_map
{'R': {'A', 'G'},
 'Y': {'C', 'T'},
 'M': {'A', 'C'},
 'K': {'G', 'T'},
 'W': {'A', 'T'},
 'S': {'C', 'G'},
 'B': {'C', 'G', 'T'},
 'D': {'A', 'G', 'T'},
 'H': {'A', 'C', 'T'},
 'V': {'A', 'C', 'G'},
 'N': {'A', 'C', 'G', 'T'}}
>>> seq.degenerate_map = 'xyz'
AttributeError: can't set attribute
>>> setattr(seq, 'degenerate_map', 'xyz')
AttributeError: can't set attribute
>>> seq.degenerate_map['A'] = 'xyz'  # this won't make any effect
>>> seq.degenerate_map  # the same
{'R': {'A', 'G'},
 'Y': {'C', 'T'},
 'M': {'A', 'C'},
 'K': {'G', 'T'},
 'W': {'A', 'T'},
 'S': {'C', 'G'},
 'B': {'C', 'G', 'T'},
 'D': {'A', 'G', 'T'},
 'H': {'A', 'C', 'T'},
 'V': {'A', 'C', 'G'},
 'N': {'A', 'C', 'G', 'T'}}
```

To really break the IUPAC codes, one needs to do `DNA.degenerate_map = 'xyz'`. But if a user really wants to change the `DNA` class itself, they should know what they are doing. Even the `@dataclass(frozen=True)` I mentioned above won't stop this (I just tested). It is like you cannot stop a user from modifying a tuple or frozenset with `_tuple = 'xyz'`.

**Comment by qiyunzhu** [comment]
Hi @xguse We have enabled subclassing the sequence classes. Thanks for your suggestion!

**Comment by clintval** [comment]
Wow, thanks everyone!! Cool to see after all these years.

Closed with #1891
